pkg/cache/multi_namespace_cache.go

```diff
@@ -37,39 +39,88 @@ type NewCacheFunc func(config *rest.Config, opts Options) (Cache, error)
+// objects. If it is undefined, a default cache will be created using the New function.
+func BuilderByNamespace(byNamespaceOpts ByNamespaceOptions) NewCacheFunc {
 	return func(config *rest.Config, opts Options) (Cache, error) {
 		opts, err := defaultOpts(config, opts)
 		if err != nil {
 			return nil, err
 		}
 
-		caches := map[string]Cache{}
-
-		// create a cache for cluster scoped resources
-		gCache, err := New(config, opts)
+		if byNamespaceOpts.NewClusterCache == nil {
+			byNamespaceOpts.NewClusterCache = New
+		}
+		clusterCache, err := byNamespaceOpts.NewClusterCache(config, opts)
 		if err != nil {
-			return nil, fmt.Errorf("error creating global cache: %w", err)
+			return nil, err
 		}
 
-		for _, ns := range namespaces {
+		nsToCache := map[string]Cache{}
+		if byNamespaceOpts.NewDefaultNamespaceCache != nil {
+			defaultNamespaceCache, err := byNamespaceOpts.NewDefaultNamespaceCache(config, ignoreNamespaces(opts, byNamespaceOpts.NewNamespaceCaches))
+			if err != nil {
+				return nil, err
+			}
+			nsToCache[corev1.NamespaceAll] = defaultNamespaceCache
+		}
+
+		for ns, newCacheFunc := range byNamespaceOpts.NewNamespaceCaches {
 			opts.Namespace = ns
-			c, err := New(config, opts)
+			nsToCache[ns], err = newCacheFunc(config, opts)
 			if err != nil {
 				return nil, err
 			}
-			caches[ns] = c
 		}
-		return &multiNamespaceCache{namespaceToCache: caches, Scheme: opts.Scheme, RESTMapper: opts.Mapper, clusterCache: gCache}, nil
+
+		return &multiNamespaceCache{
+			namespaceToCache: nsToCache,
+			clusterCache:     clusterCache,
+			RESTMapper:       opts.Mapper,
+			Scheme:           opts.Scheme,
+		}, nil
 	}
 }
 
+func ignoreNamespaces(opts Options, newObjectCaches map[string]NewCacheFunc) Options {
+	fieldSelectors := []fields.Selector{}
+	if opts.DefaultSelector.Field != nil {
+		fieldSelectors = append(fieldSelectors, opts.DefaultSelector.Field)
+	}
+	for ns := range newObjectCaches {
+		fieldSelectors = append(fieldSelectors, fields.OneTermNotEqualSelector("metadata.namespace", ns))
+	}
+	opts.DefaultSelector.Field = fields.AndSelectors(fieldSelectors...)
+	return opts
+}
+
+// MultiNamespacedCacheBuilder - Builder function to create a new multi-namespaced cache.
+// This will scope the cache to a list of namespaces. Listing for all namespaces
+// will list for all the namespaces that this knows about. By default this will create
+// a global cache for cluster scoped resource. Note that this is not intended
+// to be used for excluding namespaces, this is better done via a Predicate. Also note that
+// you may face performance issues when using this with a high number of namespaces.
+func MultiNamespacedCacheBuilder(namespaces []string) NewCacheFunc {
+	byNamespaceOpts := ByNamespaceOptions{NewNamespaceCaches: map[string]NewCacheFunc{}}
+	for _, ns := range namespaces {
+		byNamespaceOpts.NewNamespaceCaches[ns] = New
+	}
+	return BuilderByNamespace(byNamespaceOpts)
+}
```

Comment:
Now, MultiNamespacedCacheBuilder is simply a specific configuration of the `BuilderByNamespace` where the catch-all namespace-scoped cached is left as `nil` (to avoid caching namespace-scoped objects outside the specified namespaces)

Reply:
I agree with @alvaroaleman on making the functionality look similar across both options. But the question is if `BuilderByNamespace` and `MultiNamespaceCacheBuilder` support specifying cache funcs per namespace/cluster/all other namespaces wise, then what would be the underlying difference between them. Can we embed `byNamespaceOpts` directly into `MultiNamespacedCacheBuilder`? This way, the user can either provide the list of namespaces they want to cache or cache funcs for each of them. One option for this would be expose these options through `WithNamespace` or `WithCacheOptions` methods. This is definitely a major breaking change, but I guess its worth exploring instead of having two methods do the same with difference just being the parameters passed?

Reply:
I just felt it didn't make sense to break `MultiNamespacedCacheBuilder` for no good reason. I'd be on board with deprecating it though.

Reply:
Instead of having `BuilderByNamespace` and `MultiNamespaceCacheBuilder` as it is currently, can we introduce `MultiNamespaceCacheBuilderWithOptions` (or some better name) that allows both options - providing a list of namespaces, and/or cache funcs (with builder opts) and then deprecate `MultiNamespacedCacheBuilder` in following releases?

Reply:
OK updated to:
- Call the new function `MultiNamespacedCacheWithOptionsBuilder`
- Use functional options instead of a struct
- Deprecate `MultiNamespacedCacheBuilder`
- Include an example in the `MultiNamespacedCacheBuilder` deprecation message that explains how to achieve the existing functionality with the new builder function.

Does this get closer to what y'all are imagining?